### PR TITLE
Making targeting keys configurable

### PR DESCRIPTION
**Pull request by pm-harshad-mane**
## Type of change
- [x] Feature
- [x] Refactoring (no functional changes, no api changes)

## Description of change
As of now the targeting keys in Prebid.js are hard-coded, starting with "hb_" , I have made changes to make these key names configurable. So now publishers will be able to change key-names.
This will help the wrapper solutions which are build using Prebid.

### CODE REVIEW

**Comment by pm-harshad-mane** [comment]
Hello @jsnellbaker , 
Can you help me with this PR?
There are some conflicts we need to merge.

File: test/spec/unit/pbjs_api_spec.js [hunk 1]
```diff
@@ -211,20 +211,20 @@ describe('Unit: Prebid Module', function () {
...
         }
       };
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.SIZE] = '300x250';
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.AD_ID] = '233bcbee889d46d';
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.SIZE] = '728x90';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.AD_ID] = '24bd938435ec3fc';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
```
Comment (harpere) [inline test/spec/unit/pbjs_api_spec.js]: cleaner to use "computed" properties.

File: test/spec/unit/pbjs_api_spec.js [hunk 2]
```diff
@@ -258,21 +258,21 @@ describe('Unit: Prebid Module', function () {
...
```
Comment (harpere) [inline test/spec/unit/pbjs_api_spec.js]: same

File: test/spec/unit/pbjs_api_spec.js [hunk 3]
```diff
@@ -317,22 +317,21 @@ describe('Unit: Prebid Module', function () {
...
       };
-
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.SIZE] = '300x250';
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.AD_ID] = '233bcbee889d46d';
+      expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.SIZE] = '728x90';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.AD_ID] = '24bd938435ec3fc';
+      expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
```
Comment (harpere) [inline test/spec/unit/pbjs_api_spec.js]: same

File: test/spec/unit/pbjs_api_spec.js [hunk 4]
```diff
@@ -1523,21 +1522,20 @@ describe('Unit: Prebid Module', function () {
...
+        expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.SIZE] = '300x250';
+        expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+        expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.AD_ID] = '233bcbee889d46d';
+        expected['/19968336/header-bid-tag-0'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
+        
+        expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
+        expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.AD_ID] = '24bd938435ec3fc';
+        expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+        expected['/19968336/header-bid-tag1'][CONSTANTS.TARGETING_KEYS.SIZE] = '728x90';
+        expected['/19968336/header-bid-tag1']['foobar'] = '728x90';
```
Comment (harpere) [inline test/spec/unit/pbjs_api_spec.js]: same

File: test/spec/unit/pbjs_api_spec.js [hunk 5]
```diff
@@ -1790,16 +1788,16 @@ describe('Unit: Prebid Module', function () {
...
+          'adserverTargeting': {            
+            'foobar': '300x250'
           }
         }
       ];
+      $$PREBID_GLOBAL$$._bidsReceived[0]['adserverTargeting'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'appnexus';
+      $$PREBID_GLOBAL$$._bidsReceived[0]['adserverTargeting'][CONSTANTS.TARGETING_KEYS.AD_ID] = '233bcbee889d46d';
+      $$PREBID_GLOBAL$$._bidsReceived[0]['adserverTargeting'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '10.00';
+      $$PREBID_GLOBAL$$._bidsReceived[0]['adserverTargeting'][CONSTANTS.TARGETING_KEYS.SIZE] = '300x250';
+      $$PREBID_GLOBAL$$._bidsReceived[0]['adserverTargeting'][CONSTANTS.TARGETING_KEYS.DEAL + '_appnexusDummyName'] = '1234';
```
Comment (harpere) [inline test/spec/unit/pbjs_api_spec.js]: same

File: src/targeting.js [hunk 6]
```diff
@@ -15,6 +15,10 @@ export const RENDERED = 'rendered';
 const MAX_DFP_KEYLENGTH = 20;
 const TTL_BUFFER = 1000;
 
+export const TARGETING_KEYS = Object.keys(CONSTANTS.TARGETING_KEYS).map(
+  key => CONSTANTS.TARGETING_KEYS[key]
+);
+
```
Comment (harpere) [inline src/targeting.js]: same - Why not just use `Object.assign({}, CONSTANTS.TARGETING_KEYS);`?

File: src/targeting.js [hunk 7]
```diff
@@ -260,7 +264,7 @@ export function newTargeting(auctionManager) {
             winner.sendStandardTargeting ||
             standardKeys.indexOf(key) === -1)
           .map(key => ({
-            [(key === 'hb_deal') ? `${key}_${winner.bidderCode}`.substring(0, MAX_DFP_KEYLENGTH) : key.substring(0, MAX_DFP_KEYLENGTH)]: [winner.adserverTargeting[key]]
+            [(key === CONSTANTS.TARGETING_KEYS.DEAL) ? `${key}_${winner.bidderCode}`.substring(0, MAX_DFP_KEYLENGTH) : key.substring(0, MAX_DFP_KEYLENGTH)]: [winner.adserverTargeting[key]]
```
Comment (harpere) [inline src/targeting.js]: don't need the `CONSTANTS.`

File: src/native.js [hunk 8]
```diff
@@ -1,20 +1,12 @@
...
-  clickUrl: 'hb_native_linkurl',
-  cta: 'hb_native_cta',
-};
+export const nativeAdapters = [];
 
-export const NATIVE_TARGETING_KEYS = Object.keys(NATIVE_KEYS).map(
-  key => NATIVE_KEYS[key]
+export const NATIVE_TARGETING_KEYS = Object.keys(CONSTANTS.NATIVE_KEYS).map(
+  key => CONSTANTS.NATIVE_KEYS[key]
 );
```
Comment (harpere) [inline src/native.js]: unless I'm missing something, this is just copying `CONSTANTS.NATIVE_KEYS`.  Why not just use `Object.assign({}, CONSTANTS.NATIVE_KEYS);` instead?

File: test/spec/unit/core/targeting_spec.js [hunk 9]
```diff
@@ -27,15 +27,15 @@ const bid1 = {
...
-    'hb_pb': '0.53',
     'foobar': '300x250'
   },
   'netRevenue': true,
   'currency': 'USD',
   'ttl': 300
 };
+bid1['adserverTargeting'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'rubicon';
+bid1['adserverTargeting'][CONSTANTS.TARGETING_KEYS.AD_ID] = '148018fe5e';
+bid1['adserverTargeting'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '0.53';
```
[the remark this answers is not in the document]
Comment (pm-harshad-mane) [reply test/spec/unit/core/targeting_spec.js]: Done, thanks for suggestion :)

File: test/spec/unit/core/targeting_spec.js [hunk 10]
```diff
@@ -56,16 +56,16 @@ const bid2 = {
...
+  'adserverTargeting': {    
     'foobar': '300x250'
   },
   'netRevenue': true,
   'currency': 'USD',
   'ttl': 300
 };
+bid2['adserverTargeting'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'rubicon';
+bid2['adserverTargeting'][CONSTANTS.TARGETING_KEYS.AD_ID] = '5454545';
+bid2['adserverTargeting'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '0.25';
```
[the remark this answers is not in the document]
Comment (pm-harshad-mane) [reply test/spec/unit/core/targeting_spec.js]: Done, thanks for suggestion

File: test/spec/unit/core/targeting_spec.js [hunk 11]
```diff
@@ -87,15 +87,15 @@ const bid3 = {
...
-    'hb_pb': '0.75',
     'foobar': '300x600'
   },
   'netRevenue': true,
   'currency': 'USD',
   'ttl': 300
 };
+bid3['adserverTargeting'][CONSTANTS.TARGETING_KEYS.BIDDER] = 'rubicon';
+bid3['adserverTargeting'][CONSTANTS.TARGETING_KEYS.AD_ID] = '48747745';
+bid3['adserverTargeting'][CONSTANTS.TARGETING_KEYS.PRICE_BUCKET] = '0.75';
```
[the remark this answers is not in the document]
Comment (pm-harshad-mane) [reply test/spec/unit/core/targeting_spec.js]: Done, thanks for the suggestion

File: test/spec/unit/pbjs_api_spec.js (hunk 2 above)
Comment (pm-harshad-mane) [reply test/spec/unit/pbjs_api_spec.js]: Done, thanks for the suggestion

File: test/spec/unit/pbjs_api_spec.js (hunk 3 above)
Comment (pm-harshad-mane) [reply test/spec/unit/pbjs_api_spec.js]: Done, thanks for the suggestion

File: test/spec/unit/pbjs_api_spec.js (hunk 4 above)
Comment (pm-harshad-mane) [reply test/spec/unit/pbjs_api_spec.js]: Done, thanks for the suggestion

File: test/spec/unit/pbjs_api_spec.js (hunk 5 above)
Comment (pm-harshad-mane) [reply test/spec/unit/pbjs_api_spec.js]: Done, thanks for the suggestion

File: src/targeting.js (hunk 7 above)
Comment (pm-harshad-mane) [reply src/targeting.js]: TARGETING_KEYS is an array, does not have a DEAL key.
I think we need to keep this code as it is.

File: test/spec/unit/pbjs_api_spec.js [hunk 12]
```diff
@@ -720,7 +719,7 @@ describe('Unit: Prebid Module', function () {
       ajaxStub.restore();
     });
 
-    it('should get correct hb_pb with cpm between 0 - 5', function () {
+    it('should get correct ' + CONSTANTS.TARGETING_KEYS.PRICE_BUCKET + ' with cpm between 0 - 5', () => {
```
Comment (jsnellbaker) [inline test/spec/unit/pbjs_api_spec.js]: Can you change back the various `() => {` references to the `, function() {` variant?  These were implemented to make the unit tests more compliant with Mocha.

**Comment by pm-harshad-mane** [comment]
Hello @harpere and @jsnellbaker ,
Am I missing anything? is something pending from my side?

File: test/spec/unit/pbjs_api_spec.js (hunk 12 above)
Comment (pm-harshad-mane) [reply test/spec/unit/pbjs_api_spec.js]: Hello @jsnellbaker , 
sure, let me check.

Comment (pm-harshad-mane) [reply test/spec/unit/pbjs_api_spec.js]: Hello @jsnellbaker ,
I have made these changes.

**Pull request merged**